File: src/main/java/org/example/PartitionerProducer.java
package org.example;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Properties;
import java.util.concurrent.ExecutionException;

public class PartitionerProducer {

    private static final String TOPIC = "partitioner-topic-test";
    private static final String KEY = "partition_1";
    private static final String VALUE = "Zapisane na 1 partycji...";

    public static void main(String[] args) throws InterruptedException, ExecutionException {
        sendMessageToKafkaTopic();
    }

    private static void sendMessageToKafkaTopic() throws InterruptedException, ExecutionException {
        // create producer
        try (Producer<String, String> producer = createProducer()) {

            // prepare record to send
            ProducerRecord<String, String> producerRecord = new ProducerRecord<>(
                    TOPIC,
                    KEY,
                    VALUE
            );

            // send action
            System.out.println(producer.send(producerRecord).get());
        }
    }

    private static Producer<String, String> createProducer() {
        Properties properties = new Properties();

        properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
        properties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        properties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        properties.put(ProducerConfig.CLIENT_ID_CONFIG, "out-producer");

        // Nasz partitioner
        properties.put(ProducerConfig.PARTITIONER_CLASS_CONFIG, OurPartitioner.class);

        return new KafkaProducer<>(properties);
    }

}

File: python/PartitionerProducer.py
from confluent_kafka import Producer
from confluent_kafka.serialization import StringSerializer


class PartitionerProducer:

    __TOPIC = "partitioner-topic-test"
    __KEY = "partition_1"
    __VALUE = "Zapisane na 1 partycji..."

    @staticmethod
    def send_message_to_kafka_topic():
        # create producer
        producer = PartitionerProducer.create_producer()

        # prepare serializer (not in producer config!!!)
        string_serializer = StringSerializer('utf_8')

        # simulate partitioner object behavior
        partition_no = PartitionerProducer.get_partition_no_from_key(PartitionerProducer.__KEY)

        # prepare record to send and send action
        producer.poll(0)
        producer.produce(
            topic=PartitionerProducer.__TOPIC,
            key=string_serializer(PartitionerProducer.__KEY),
            value=string_serializer(PartitionerProducer.__VALUE),
            partition=partition_no
        )
        producer.flush()

    @staticmethod
    def create_producer() -> Producer:
        # this producer config map is not equivalent ProducerConfig.java :|
        return Producer(
            {
                'bootstrap.servers': 'localhost:9092',
                'client.id': 'our-producer',
            }
        )

    @staticmethod
    def get_partition_no_from_key(key):
        # it's equivalent OurPartitioner.java behavior
        return int(key.split("_")[-1])


def main():
    PartitionerProducer.send_message_to_kafka_topic()


if __name__ == '__main__':
    main()
